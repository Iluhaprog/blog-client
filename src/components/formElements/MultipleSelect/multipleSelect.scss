@import '../../../vars.scss';

.multiple-select {
    position: relative;
    outline: none;
    min-width: 300px;
    min-height: 50px;
    width: 100%;
    height: 100%;
    padding: 10px;
    border: $border-2;
    border-radius: 5px;
    transition: 0.3s ease-in-out;
    
    .values {
        list-style: none;
        margin: 0;
        padding: 0;

        .tag {
            font-size: 20px;
            margin-bottom: 10px;
        }
    }
    .options {
        position: absolute;
        top: 100%;
        left: -1px;
        width: calc(100% + 3px);
        height: 0;
        overflow: hidden;
        transition: .3s ease-in-out;
        background-color: #fff;
        border-radius: 5px;
        border: $border-2;
        border-top: 0;
    }
    
    .option {
        padding: 10px 15px;
        border-bottom: $border-2;
        transition: .15s ease-in-out;

        &:last-child {
            border-bottom: 0;
        }

        &:hover {
            background-color: $text-color-5;
        }
    }
    
    &:focus {
        border: $border-2;
        border-bottom: 0;
        box-shadow: $shadow;
        height: calc(auto + 200px);
        border-radius: 5px 5px 0 0;
        
        .options {
            height: 200px;
            overflow-y: scroll;
            border-radius: 0 0 5px 5px;
        }
    }
}