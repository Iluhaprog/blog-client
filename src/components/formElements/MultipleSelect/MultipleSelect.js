import React from 'react';
import PropTypes from 'prop-types';
import { TagBox } from '../../boxes';
import { Row } from '../../containers';
import './multipleSelect.scss';

class MultipleSelect extends React.Component {
    constructor(props) {
        super(props);
        this.select = React.createRef();
        this.focus = this.focus.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.state = {
            tags: [],
        }
    }

    focus() {
        this.select && this.select.current.focus();
    }

    componentDidMount() {
        this.setState({
            tags: this.props.values
        });
    }

    handleChange(value) {
        const el = this.state.tags.filter(option => option.id === value.id);
        if (!el.length) {
            this.setState({
                tags: [...this.state.tags, value],
            })
            this.props.onChange(value);
        }
    }
    
    render() {
        const { options } = this.props;

        return (
            <div 
                className='multiple-select'
                ref={this.select}
                onClick={this.focus}
                tabIndex='0'
            >
                <ul className='values'>
                    <Row justifyContent='fs' alignItems='c' wrap='w'>
                        {
                            this.state.tags.map(value => (
                                <TagBox 
                                    key={value.id}
                                    title={value.title + ''} 
                                />
                            ))
                        }
                    </Row>
                </ul>
                <div className='options'>
                    {
                        options.map(option => (
                            <div 
                                key={option.id}
                                className='option'
                                onClick={() => {this.handleChange(option)}}
                            >
                                {option.title}
                            </div>
                        ))      
                    }
                </div>
            </div>
        );
    }
}

MultipleSelect.defaultProp = {
    options: [],
    values: [],
};

MultipleSelect.propTypes = {
    options: PropTypes.array,
    values: PropTypes.array,
};

export default MultipleSelect;