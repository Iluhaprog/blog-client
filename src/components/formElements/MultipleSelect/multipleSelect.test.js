import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react';
import MultipleSelect from './MultipleSelect';

describe('Test view of multiple select', () => {
    test('Without values and options', () => {
        const { asFragment } = render(
            <MultipleSelect
                values={[]}
                options={[]}
                onChange={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
    test('With values and options', () => {
        const { asFragment } = render(
            <MultipleSelect
                values={[{ id: 1, title: 11 }]}
                options={[{ id: 1, title: 'test1' }, { id: 2, title: 'test2' }]}
                onChange={() => {}}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Test events of multiple select', () => {
    test('Should handle change', async () => {
        const handleChange = jest.fn();
        const { container } = render(
            <MultipleSelect
                values={[]}
                options={[{ id: 1, title: 'test1' }, { id: 2, title: 'test2' }]}
                onChange={handleChange}
            />
        );
        fireEvent.click(container.querySelector('.option'));
        await waitFor(() => {
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(handleChange).toHaveBeenCalledWith({ id: 1, title: 'test1' });
        });      
    });
});