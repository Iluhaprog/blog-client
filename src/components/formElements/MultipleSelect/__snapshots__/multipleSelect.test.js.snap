// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test view of multiple select With values and options 1`] = `
<DocumentFragment>
  <div
    class="multiple-select"
    tabindex="0"
  >
    <ul
      class="values"
    >
      <div
        class="row row_jc-fs row_ai-c row_w-w"
      >
        <li
          class="tag tag__"
        >
          11
        </li>
      </div>
    </ul>
    <div
      class="options"
    >
      <div
        class="option"
      >
        test1
      </div>
      <div
        class="option"
      >
        test2
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Test view of multiple select Without values and options 1`] = `
<DocumentFragment>
  <div
    class="multiple-select"
    tabindex="0"
  >
    <ul
      class="values"
    >
      <div
        class="row row_jc-fs row_ai-c row_w-w"
      />
    </ul>
    <div
      class="options"
    />
  </div>
</DocumentFragment>
`;
